ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/BlockPropagationManager.java

```diff
@@ -155,29 +155,49 @@ private void clearListeners() {
   private void onBlockAdded(final BlockAddedEvent blockAddedEvent) {
     // Check to see if any of our pending blocks are now ready for import
     final Block newBlock = blockAddedEvent.getBlock();
+    traceLambda(
+        LOG,
+        "Block added event type {} for block {}. Current status {}",
+        blockAddedEvent::getEventType,
+        newBlock::toLogString,
+        () -> this);
 
+    // If there is no children to process, maybe try non announced blocks
+    if (!maybeProcessPendingChildrenBlocks(newBlock)) {
+      traceLambda(
+          LOG, "There are no pending blocks ready to import for block {}", newBlock::toLogString);
+      maybeProcessNonAnnouncedBlocks(newBlock);
+    }
+
+    if (blockAddedEvent.getEventType().equals(EventType.HEAD_ADVANCED)) {
+      final long head = protocolContext.getBlockchain().getChainHeadBlockNumber();
+      final long cutoff = head + config.getBlockPropagationRange().lowerEndpoint();
+      pendingBlocksManager.purgeBlocksOlderThan(cutoff);
```

Comment:
Seems suspicious that previously access to this was guarded by a `synchronized (pendingBlocksManager)` but this access is not.

Reply:
This PR does not change that. It was previously accessed outside of the `synchronized` block.

Reply:
hmm, not related to this PR then but I suspect there's a race condition there because the prune winds up trying to update both `pendingBlocks` and `pendingBlocksByParentHash`. If `PendingBlocksManager.registerPendingBlock` and `deregisterPendingBlock` wind up running at the same time, the purge may run in between register adding to `pendingBlocks` and adding to `pendingBlocksByParentHash`. The purge would then only remove from `pendingBlocks` and we'd wind up with an entry in `pendingBlocksByParentHash` that never gets removed.  It's very unlikely and a very small memory leak with no effect on functionality though.

Plus there may be some other restriction somewhere I'm missing that ensures those two methods never run at the same time for the same block.